lib/rails_state_machine/state_machine.rb

```diff
@@ -74,11 +82,24 @@ def model_module_eval(&block)
       @model_module.module_eval(&block)
     end
 
+    def check_if_states_already_defined
+      @states_by_name.each do |state_name, _|
+        if other_state_machines.any? { |state_machine| (state_machine.has_state?(state_name)) && (state_machine.prefix == prefix) }
+          raise(StateAlreadyDefinedError, "State :#{state_name} has already been defined in another state_machine. Please use \"prefix\" as a keyword argument.")
```

Comment:
Maybe this is a bit more explicit?

```suggestion
          raise(StateAlreadyDefinedError, "State #{state_name.inspect} has already been defined in the #{state_machine.state_attribute.inspect} state machine. You may use the :prefix option when defining a state machine to avoid that.")
```

(would require `state_attribute` to be declared as an `attr_reader`)

Reply:
> where does `state_machine` come from? `state_machine` is only defined in the `#any`-Block itself.

Since `raise` stops execution, you should just enumerate with `each` instead of `any?` and `raise` from inside the block.

```ruby
    def check_if_states_already_defined
      @states_by_name.each do |state_name, _|
        other_state_machines.each do |state_machine|
          if state_machine.has_state?(state_name) && state_machine.prefix == prefix
            raise StateAlreadyDefinedError, "State #{state_name.inspect} has already been defined in the #{state_machine.state_attribute.inspect} state machine. You may use the :prefix option when defining a state machine to avoid that."
          end
        end
      end
    end
```

Also, side not to your suggested change: if you care about only the first match, use [`find`/`detect`](https://ruby-doc.org/3.2.2/Enumerable.html#method-i-detect) instead of `select`.

So if you want to take that approach, you should do something like this:

```ruby
    def check_if_states_already_defined
      @states_by_name.each do |state_name, _|
        conflicting_state_machine = other_state_machines.detect { |state_machine|state_machine.has_state?(state_name) && state_machine.prefix == prefix }
        if conflicting_state_machine
          raise StateAlreadyDefinedError, "State #{state_name.inspect} has already been defined in the #{conflicting_state_machine.state_attribute.inspect} state machine. You may use the :prefix option when defining a state machine to avoid that."
        end
      end
    end
```

Reply:
I will take your first suggestion. Thank you for the tip below!